### ISSUE TITLE
Testing behavior for non-standard ports.

### ISSUE THREAD

**Issue by mikewest** [w3c/web-platform-tests#4751]
https://github.com/whatwg/fetch/pull/464 and https://github.com/whatwg/fetch/issues/482 are proposals to change Fetch's behavior to block a few non-standard ports. We don't have test servers running on these ports, which means that for most things, the failure case (network error because of a timeout) will be the same as the success case (network error due to blockage) in terms of its web-visible side-effects.

It's possible, I suppose to run a timing attack as a test, and to call it "fail" if it takes more than Xms, but that feels a bit icky.

In https://github.com/whatwg/fetch/issues/482#issuecomment-277956257, @annevk suggested simply documenting behaviors that we can't reasonably test in WPT somewhere.

@jgraham, @ms2ger, @foolip, thoughts?

**Comment by jakearchibald** [comment]
Could we just add some servers to those ports? I'm not massively familiar with wpt so apologies if that's naive.

**Comment by mikewest** [comment]
Most of the ports are below 1024, which means we'd need root access to map them. I'm not thrilled about giving wptserve root access on my development box, but it's certainly something we could consider for the "real" server. I guess there's also potential for conflict, since many of these ports represent services that might actually be running on your machine (`ssh`, for instance, is something I'd be sad about giving up locally when running WPT).

**Comment by asutherland** [comment]
If documenting, perhaps use a JSON file for the canonical list so that browsers can specifically test using non-wpt tests?  For example, Gecko exposes [bool nsIIOService::allowPort(port, scheme)](http://searchfox.org/mozilla-central/source/netwerk/base/nsIIOService.idl#163) which would allow an automated test to help enforce consistency.

The other testable alternative would be to introduce a backdoor mechanism like Gecko's Mochitest's [SpecialPowers](https://developer.mozilla.org/en-US/docs/SpecialPowers) hook that lets content-privileged tests ask for privileged things or white/glass-box style assertions to be done on its behalf.  That's probably a very slippery slope, though, since it would be tempting for specific browser engines to introduce allegedly-but-not-actually-generic uses of the mechanism which might result in WPT tests that don't actually check anything if the backdoor helper isn't implemented.

**Comment by mikewest** [comment]
@asutherland: I think that a JSON file makes sense for whatwg/fetch#482's changes to the list of bad ports, but makes less sense for whatwg/fetch#464's change to the behavior of subresources with a scheme of `ftp`. It's not really clear how we'd go about _really_ testing that without spinning up an FTP server, which I don't think anyone wants to do.

**Comment by asutherland** [comment]
@mikewest Agreed that the JSON approach doesn't work or particularly make sense for the http-fetching-ftp https://github.com/whatwg/fetch/pull/464 case because the decision is likely implemented deep enough inside fetch to be unreachable.  Maybe an optional mechanism to annotate the expected errors with the secret "why" can be added that browser test runners can consume and map to their own internal affordances?  Disclaimer: Unlikely to be a small undertaking.

For example, for fetch encountering a redirect with { redirect: "error" } there's the WPT test https://github.com/w3c/web-platform-tests/blob/master/fetch/api/redirect/redirect-mode.js#L15 which does `promise_rejects(test, new TypeError(), fetch(url + urlParameters, requestInit));`.

promise_rejects could be made to take an additional argument that's a simple JSON-serializable object.  In the case of the redirect error we could pass { errorSecretType: "fetch:redirect-converted-to-error" }.  For the http-fetching-ftp case we'd pass { errorSecretType: "fetch:forbidden-ftp-transition" }.  For bad ports { errorSecretType: "fetch:bad-port" }.

The error objects and strings are arbitrary but we'd hopefully adopt a consistent convention.  The point is just to associate the meta-data with the error result.  It's on the browser-specific test executor to find some way to get the data out and check it.  A new requirement would be that the executors would need to be able to interpolate some code into testharness.js.

For example, for Firefox it would arguably be nice for its devtools if they could improve on the generic string "NetworkError when attempting to fetch resource." that is all the information its TypeError provides for fetch network errors at this time.  To support this devtools goal, Gecko could store a hidden error code on its InternalResponse rep and propagate it to its internal ErrorResult type and onwards to SpiderMonkey's JSErrorReport type.  The JS debugger API, never available to content, can then also be taught how to expose the secret type of the Error.  This provides the devtools enhancement.

Then the extra checking is built on top of that.  The testharness.js monkeypatch would log the rejected exception via console.log() along with the provided { errorSecretType } meta-data.  Gecko's Console API uses a custom structured clone implementation which can be made to serialize the uncloneable Error and its secret type.  (This matters for workers, not main-thread console calls.  Anything uncloneable gets coerced to a string by the custom write handler right now.)    Finally, testharness_marionette.js adds an observer on console log messages and performs assertion checking on the log entries as they are generated, only spinning up the debugger infrastructure as needed.  This logic would also be responsible for having a mapping between the arbitrary "fetch:bad-port" and however Gecko encodes the secret error.

The standard of "tests provide metadata" and "your browser's test executor provides massive hacks to enforce them" would also allow other options like running regexps against internal developer-only logging mechanism, etc.  (For example, in Gecko/Necko, our HTTP implementation can be dynamically configured to emit an extremely detailed amount of debug information via MOZ_LOG that the tests could process to assert specific edge-cases have been triggered.)

**Comment by foolip** [comment]
There are woefully many things that are currently web exposed but not possible to write automated tests for using web-platform-tests, and in this case not even manual tests unless one asks the user to do certain things as root while the test is waiting, which would be silly.

@jgraham, is there already a label we can use to keep track of things that are untestable, as opposed to just lacking tests? Maybe add `type:untestable` or `type:needs-automation`?

**Comment by foolip** [comment]
@mikewest @jakearchibald are there tests in Chromium which aren't shared because of this? Adding the priority:roadmap label on the assumption that there is, so that we don't lose track of this. I have no ETA for when someone would work on this though. @gsnedders, is it as simple has having wptserve listen to some more ports?

**Comment by mikewest** [comment]
I don't think we have any layout tests we could share through WPT. Tests like https://cs.chromium.org/chromium/src/third_party/WebKit/LayoutTests/http/tests/security/deprecated-subresource-requests-expected.txt?type=cs&q=http/tests/security/deprecated-subresource-requests-expected.txt&l=1 rely on console messages to verify behavior, as for pretty much all of these ports, the success case is identical to the failure case (as we don't have a server listening on most of the ports).

I think I'm opposed to the idea of requiring root access to run the WPT server on developer machines to get access to ports below 1024, though. The benefit of getting coverage for this very specific restriction seems quite outweighed by the risks that access would entail to folks running the testing environment.

**Comment by foolip** [comment]
Thanks @mikewest, demoting this to backlog based on that. Let's fix things that are blocking shared interop tests first :)

**Comment by annevk** [comment]
Note that as I discovered in https://github.com/web-platform-tests/wpt/pull/11249 we have a two "bad port" tests. They only test for failure however, not that other ports do work.

**Comment by stalgiag** [comment]
Reviving this as I believe I have found an approach that wasn't available when this thread was last active that satisfies the main concerns here. #62882 routes the fetches in request-bad-port.any.js through the per-test PAC support introduced in #34145. With this, a request that isn't blocked reaches wptserve even without anything listening on that port. This allows us to test the blocking for every port on the list, even those below 1024, without root or new listeners, and a control subtest checks that a port that isn't bad does work, per annevk's point above.

Using this, I caught that Chrome is missing blocks for 4190 and 6679 (https://issues.chromium.org/issues/564770251). Safari's WebDriver doesn't support PAC so there only the control subtest fails.

**Comment by annevk** [comment]
Oh, that's nice! Your Chrome issue is a duplicate of https://issues.chromium.org/issues/332391867.

**Comment by mikewest** [comment]
Cool! Thanks for digging into this, @stalgiag. :)

/cc @ricea for the Chromium bits.
